ext/standard/ftp_fopen_wrapper.c

```diff
@@ -720,18 +720,17 @@ php_stream * php_stream_ftp_opendir(php_stream_wrapper *wrapper, const char *pat
 	if (hoststart == NULL) {
 		hoststart = resource->host;
 	}
-	datastream = php_stream_sock_open_host(hoststart, portno, SOCK_STREAM, 0, 0);
-	if (datastream == NULL) {
-		goto opendir_errexit;
-	}
 
 	result = GET_FTP_RESULT(stream);
 	if (result != 150 && result != 125) {
 		/* Could not retrieve or send the file
 		 * this data will only be sent to us after connection on the data port was initiated.
 		 */
```

Comment:
I suspect this code was copied from php_stream_url_wrap_ftp() where it might make a bit more sense.

My interpretation that data channel should not be opened is based on section 6 (state diagrams) of the RFC 959. Response with 4xx or 5xx would lead into F-state (failure) without attempting to open data channel. 1xx response (e.g. 125 or 150) should lead to opening the data channel, processing as required, and return to W-state.

This also matches what I see in real life. IIS 7.5 and 8.5 both respond with 550. ProFTPD responds with 450 but may be configured to respond with 226 as well. At this point server will not accept passive data stream.

It is OS and/or firewall dependent whether to RST or silently ignore the attempts to open data channel without server listening for it.


Further, datastream error handling `if (datastream == NULL) {` will incorrectly leave last good response, usually "200 Type set to A", in tmp_line which will then show as in Warning:

```
PHP Warning:  opendir(): connect() failed: Network is unreachable in test.php on line 2
PHP Warning:  opendir(ftp://example.com/asd): failed to open dir: FTP server reports 200 Type set to A.
 in test.php on line 2
```

Handling of the error condition from php_fopen_do_pasv() seems to be suffering from the same problem, but I digress.